matrix-api-r0.3.0.el

```diff
@@ -0,0 +1,480 @@
+
+(cl-defmethod matrix-sync-rooms ((session matrix-session) rooms)
+  "Process ROOMS from sync response on SESSION."
+  ;; https://matrix.org/docs/spec/client_server/r0.3.0.html#id167
+  (cl-loop for room in rooms
+           always (pcase room
+                    (`(join . ,_) (matrix-sync-join session room))
+                    (`(invite .  ,_) (matrix-log "Would process room invites: %s" room))
+                    (`(leave . ,_) (matrix-log "Would process room leaves: %s" room)))))
+
+(cl-defmethod matrix-sync-join ((session matrix-session) join)
+  "Sync JOIN, a list of joined rooms, on SESSION."
+  ;; https://matrix.org/docs/spec/client_server/r0.3.0.html#id167
+  (with-slots (rooms) session
+    (cl-loop for it in (cdr join)
+             always (pcase-let* ((`(,joined-room-id . ,joined-room) it)
+                                 ;; Room IDs are decoded from JSON as symbols, so we convert to strings.
+                                 (room-id (symbol-name joined-room-id))
+                                 (params '(state timeline ephemeral account_data unread_notifications))
+                                 (room (or (--first (equal (oref it id) room-id)
+                                                    rooms)
+                                           ;; Make new room
+                                           (let ((new-room (matrix-room :session session
+                                                                        :id room-id)))
+                                             (push new-room rooms)
+                                             new-room))))
+                      (cl-loop for param in params
+                               for method = (intern (concat "matrix-sync-" (symbol-name param)))
+                               always (if (functionp method)
+                                          (funcall method room (a-get joined-room param))
+                                        ;; `warn' seems to return non-nil.  Convenient.
+                                        (warn "Unimplemented method: %s" method-name)))))))
+
+(cl-defmethod matrix-sync-state ((room matrix-room) state)
+  "Sync STATE in ROOM."
+  (with-slots (state) room
+    (pcase-let (((map events) state))
+      ;; events is an array, not a list, so we can't use --each.
+      (seq-doseq (event events)
+        (push event state)))))
+
+(cl-defmethod matrix-sync-timeline ((room matrix-room) timeline-sync)
+  "Sync TIMELINE-SYNC in ROOM."
+  (with-slots (timeline prev-batch) room
+    (pcase-let (((map events limited prev_batch) timeline-sync))
+      (seq-doseq (event events)
+        (push event timeline))
+      (setq prev-batch prev_batch))))
+
+(cl-defmethod matrix-messages ((session matrix-session) room-id
+                               &key (direction "b") limit)
+  "Request messages for ROOM-ID in SESSION.
+DIRECTION must be \"b\" (the default) or \"f\".  LIMIT is the
+maximum number of events to return (default 10)."
```

Comment:
`limit` isn't an optional argument here, so I think this docstring needs to be updated

Reply:
Actually `limit` is optional, because the API defaults it to 10, and we remove keys with null values before sending the request, and since it's a keyword arg, it can be left unspecified.